### ISSUE TITLE
Bug: Test runner spends most of the time waiting

### ISSUE THREAD

**Issue by marvinhagemeister** [denoland/deno#20438]
Noticed that we spend a lot of time seemingly doing nothing when running synchronous tests.

![Screenshot 2023-09-10 at 02 12 55](https://github.com/denoland/deno/assets/1062408/1d706985-728d-474e-95ad-bf49f18b4a2d)

Profile https://profiler.firefox.com/from-url/http%3A%2F%2F127.0.0.1%3A3000%2F0gzmz394fp09y64q0s6l1if83s7j2wdsyl4qixs%2Fprofile.json/stack-chart/?globalTrackOrder=0&hiddenLocalTracksByPid=62716-013w9&symbolServer=http%3A%2F%2F127.0.0.1%3A3000%2F0gzmz394fp09y64q0s6l1if83s7j2wdsyl4qixs&thread=0&v=10

| What | Time |
|---|--:|
| Node + mocha | 450ms |
| Bun | 402ms |
| Deno |  4s |

## Steps to reproduce

1. Create a file `generate.ts` with the following contents:
  ```tsx
  import * as path from "https://deno.land/std@0.201.0/path/mod.ts";
  
  const __dirname = new URL(".", import.meta.url).pathname;
  const dir = path.join(__dirname, "tests");

  try {
    Deno.removeSync(dir, { recursive: true });
  } catch {}
  try {
    Deno.mkdirSync(dir);
  } catch {}

  const file = path.join(__dirname, "tests", `all.test.ts`);
  let code = "";
  for (let i = 0; i < 1000; i++) {
    code += `Deno.test("Test ${i}", () => {});\n`;
  }
  Deno.writeTextFileSync(file, code);
  ```
2. Run `deno run -A generate.ts`
3. Run `deno test -A tests/all.tests.ts`

**Comment by marvinhagemeister** [comment]
Looked a bit deeper into this.

- The `sanitizeOps` is enabled by default and makes synchronous tests ~4x slower. Just disabling this alone brings down test times from `4s` -> `911ms`
- We track test/step location by creating an error and parsing the stack trace. This is super expensive and we should switch to an AST based solution like in #20435 . This brings down test times further from `911ms` -> `522ms`
- We escape each test name by invoking the iterator protocol + calling `.replaceAll` [multiple times](https://github.com/denoland/deno/blob/c228adc27d5b605286d3a9ca924e657308aebb74/cli/js/40_testing.js#L529-L534). This function is pretty inefficient too. To get an estimate, I made it a no-op function and that reduces test times from `522ms` -> `443ms`
- Skipping `sanitizeExit` and `sanitizeResources` yields another small speedup from `443ms` -> `421ms`.

**Comment by bartlomieju** [comment]
I see that I'm late to the party as you figured out most of the slow down.

> The sanitizeOps is enabled by default and makes synchronous tests ~4x slower. Just disabling this alone brings down test times from 4s -> 911ms

Op sanitization is by far the biggest problem - due to how timers work we need to do 1ms timeout to properly assert ops. We can optimize it by checking if there are any pending timers done in the test case and not do this 1ms timeout to still sanitize properly. I started doing that in #20082.

> We track test/step location by creating an error and parsing the stack trace. This is super expensive and we should switch to an AST based solution like in https://github.com/denoland/deno/pull/20435 . This brings down test times further from 911ms -> 522ms

I'm not sure this is feasible - currently locations for tests coming from `node:test` and `std/testing/bdd.ts` are off and we need a better mechanism for assinging them that is being worked on in #20047. I would be curious to hear what you'd want to do here.

> We escape each test name by invoking the iterator protocol + calling .replaceAll [multiple times](https://github.com/denoland/deno/blob/c228adc27d5b605286d3a9ca924e657308aebb74/cli/js/40_testing.js#L529-L534). This function is pretty inefficient too. To get an estimate, I made it a no-op function and that reduces test times from 522ms -> 443ms

Yikes, I added it, but honestly I didn't measure the impact. We should definitely optimize it to be cheaper (probably checking if any of the "prohibited" chars is present in the test name, before escaping it.

>  Skipping sanitizeExit and sanitizeResources yields another small speedup from 443ms -> 421ms.

These two are extremely useful and we should most likely keep them as is.

All in all, addressing op sanitizer slow down should be the first thing we tackle.

**Comment by marvinhagemeister** [comment]
> I'm not sure this is feasible - currently locations for tests coming from node:test and std/testing/bdd.ts are off and we need a better mechanism for assinging them that is being worked on in https://github.com/denoland/deno/issues/20047. I would be curious to hear what you'd want to do here.

To me parsing the stack trace seems always a little brittle. My hunch is that doing an AST prepass with SWC and grab the locations from that will be faster and more reliable too.

**Comment by bartlomieju** [comment]
@marvinhagemeister paint me skeptical, but I might turn around if I see a working prototype :D would you be interested in opening a PR for it? I'll focus on the sanitizer problem.

**Comment by piscisaureus** [comment]
> > The sanitizeOps is enabled by default and makes synchronous tests ~4x slower. Just disabling this alone brings down test times from 4s -> 911ms
> 
> Op sanitization is by far the biggest problem - due to how timers work we need to do 1ms timeout to properly assert ops. We can optimize it by checking if there are any pending timers done in the test case and not do this 1ms timeout to still sanitize properly. I started doing that in #20082.

What if you used `setImmediate()` (internally backed by [`setTimeoutUnclamped()`](https://github.com/denoland/deno/blob/4460336fdae3a6c0c8ebe8e6b8842604743cf985/ext/web/02_timers.js#L348)) rather than `setTimeout()`?

**Comment by bartlomieju** [comment]
They're all backed by the same op (`op_sleep`)

**Comment by piscisaureus** [comment]
> They're all backed by the same op (`op_sleep`)

Yes but `setTimeout(a, 1)` really means `setTimeout(a, 4)`, so if you run 1000 tests (as in @marvinhagemeister's test case) then you end up doing nothing for 4*1000 = 4 seconds.

setImmediate does not have this issue.

**Comment by bartlomieju** [comment]
Sorry I misunderstood you. I'll try that solution.

**Comment by bartlomieju** [comment]
I used `setTimeoutUnclamped` with value of 1 which should already lower the overhead. `setImmediate` won't work here because it uses `setTimeoutUnclamed` with a value of 0 which is not enough time for some scenarios (`op_sleep` and `op_worker_recv_control`). Both of these ops use "deferred" ops which have a slightly tricky timing.

**Comment by piscisaureus** [comment]
> I used `setTimeoutUnclamped` with value of 1 which should already lower the overhead. `setImmediate` won't work here because it uses `setTimeoutUnclamed` with a value of 0 which is not enough time for some scenarios (`op_sleep` and `op_worker_recv_control`). Both of these ops use "deferred" ops which have a slightly tricky timing.

Did it work? If so you can try `await new Promise(res => setImmediate(() => setImmediate(res)))` :)

**Comment by bartlomieju** [comment]
Using 4 nested `setTimeoutClamped(..., 0)` seemed to work for most tests, but is still flaky. I will think a bit more how this can be improved.

**Comment by bartlomieju** [comment]
Seems like spinning the event loop for a little bit between the tests solves this issue, but I hit #20443 along the way :)

**Comment by huix11** [comment]
d

**Comment by lucacasonato** [comment]
Fixed by #20550
